boards: arm: Add PAN1783A evaluation board

pideu-sj opened the pull request:
The PAN1783A evaluation board is a development tool for the nRF5340 from Nordic Semiconductor.

pideu-sj committed:
boards: arm: Add PAN1783A evaluation board

The PAN1783A evaluation board is a development tool for the nRF5340
from Nordic Semiconductor.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented:
Hi @anangl and @carlescufi 

you have already reviewed the evaluation board "pan1783_evb": https://github.com/zephyrproject-rtos/zephyr/pull/63104

The "pan1783**a**_evb" evaluation board is mainly a clone of the pan1783_evb, just with some modifications on the HF path. The module uses a bottom pad and a ufl connector including a external antenna instead of an chip antenna. 

**Diff pan1783a_evb <-> pan1783_evb**

- Other HF path is shortly described, 
- the designation changed in all files from "pan1783_evb" to "pan1783a_evb" and 
- another image is used for this evaluation board.

Note: A third variant of PAN1783 evaluation board will follow. The third one will have a power amplifier mounted and needs some additional configurations in devicetree. I will start the PR after this one here.

nordicjm reviewed (changes requested):
If this is just a variation then it should be in the same folder as the existing board with a different name or use the device revision to use overlay files which change what is different on this module

nordicjm commented:
> Note: A third variant of PAN1783 evaluation board will follow. The third one will have a power amplifier mounted and needs some additional configurations in devicetree. I will start the PR after this one here.

This would be fine being a different board in a separate directory

pideu-sj commented:
> If this is just a variation then it should be in the same folder as the existing board with a different name or use the device revision to use overlay files which change what is different on this module

Hi @nordicjm 
thanks for your comment. When considering uniformity for our evaluation boards for this PAN1783 Nordic module, I think we should either place all three variants in one directory or spend three separate directories. But when placing all the stuff in one folder I'm a bit afraid that it gets confusing. 
Because of the two cores the nRF5340 has already 12 files for one variant, that increasing to 36 files in one folder, when having 3 variants. So in my opinion three separate directories are clearer? 
But of course I would try to add them all to one directory, if this is the common way. After my short justification, why I did it this way, what would you say should I do?

pideu-sj commented:
Hi @nordicjm,
any thoughts to my last comment? In the meantime I already tried to get all files for the different boards into one folder. But like I mentioned above, it gets very confusing with these many files for the different cores and the different boards. So I would really prefer to spend three different folders for the different board variants. And please note, that If we would add the none secure TFM files in the future for these boards, this effect will even increase.

nordicjm commented:
The number of files shouldn't matter, people aren't going to be looking at the files, if the hardware is exactly the same except for the antenna which does not need any firmware changes, I don't really see why that needs 2 board folders.

nordicjm commented:
Actually if this is just the exact same hardware, why is a second board file needed at all when the existing board file can be used? Why does someone need to build for `pan1783a_evb` if they have the hardware, surely building for `pan1783_evb` would give you a 100% same hex file for an application?

pideu-sj commented:
> Actually if this is just the exact same hardware, why is a second board file needed at all when the existing board file can be used? Why does someone need to build for `pan1783a_evb` if they have the hardware, surely building for `pan1783_evb` would give you a 100% same hex file for an application?

You are right. Technically there will be no difference between them. But the decision was made on our site, to sell these evaluation boards as a variant with designation "pan1783**a**_evb". So if a customer tries to build his own software on this eval board variant, it should be available in the dropdown list. If there is the **pan1783_evb** and later the **pan1783a_pa_evb** available and not the **pan1783a_evb** it's also confusing for the customer.

Or is there maybe also another way in zephyr, to find two boards designations and use same files for building?

> And the link in your documentation points to the pan1783_evb website so it further confirms that this should be part of the existing board folder.

Oh thannks for pointing me to this. This is still a mistake. there is a "a" missing behind all the "pan1783" in the links. I will fix this today and do a rebase.

nordicjm commented:
I have spoken with the board/soc maintainer and they have suggested to use a board revision for this instead: https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#multiple-board-revisions

pideu-sj commented:
> I have spoken with the board/soc maintainer and they have suggested to use a board revision for this instead: https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#multiple-board-revisions

Thanks for the link, I didn't even know about this feature. I will test this and come back to you.

pideu-sj commented:
Hi @nordicjm,
any thoughts on my last comment?

anangl commented:
@pideu-sj I agree that board revision mechanism is not a good solution in this case.
I'd suggest to use the same scheme as is used in [arduino_nano_33_ble](https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/arduino_nano_33_ble), i.e. define all boards in a common folder with just separate `*_defconfig`, `*.yaml`, and `*.dts` files for all the needed targets and common parts of devicetree stuff in `*.dtsi` files. You already have such `*.dtsi` files for `pan1783_evb`, so you can include them from `*.dts` files added for the new targets. I'd just change their names (maybe just dropped the `pan1783_evb_` prefix) so that it is clear they are used for all the boards.
Would that work for you?

pideu-sj commented:
@anangl 
Thanks a lot for your review and your comment.
I've now implemented it the way you've suggested:
* I've added pan1783a files for `defconfig`, `yaml` and `dts` 
* I've renamed main folder from `pan1783_evb` to `pan1783_pan1783a_evb`
* I've removed the prefixes from shared `*.dtsi` files
* I've updated index.rst to describe and link both evb versions

Looks a bit confusing now, when looking into the main folder, because network and application core versions for `pan1783_evb` and `pan1783a_evb` are needed, but I've made some tests, and it looks like it's working fine. So I hope we can make progress with this version.

Thanks a lot!
Steffen

nordicjm commented on boards/arm/pan1783_pan1783a_evb/CMakeLists.txt:
```diff
@@ -5,3 +5,8 @@ if ((CONFIG_BOARD_PAN1783_EVB_CPUAPP) AND (CONFIG_BOARD_ENABLE_CPUNET))
 zephyr_library_sources(pan1783_evb_cpunet_reset.c)
 endif()
+
+if ((CONFIG_BOARD_PAN1783A_EVB_CPUAPP) AND (CONFIG_BOARD_ENABLE_CPUNET))
+zephyr_library()
+zephyr_library_sources(pan1783a_evb_cpunet_reset.c)
```
no spaces before brackets in cmake, and indent is 2 spaces when inside if()

nordicjm commented on boards/arm/pan1783_pan1783a_evb/board.cmake:
```diff
@@ -4,9 +4,17 @@ if(CONFIG_BOARD_PAN1783_EVB_CPUAPP)
 board_runner_args(jlink "--device=nrf5340_xxaa_net" "--speed=4000")
 endif()
 
+if(CONFIG_BOARD_PAN1783A_EVB_CPUNET)
+board_runner_args(jlink "--device=nrf5340_xxaa_net" "--speed=4000")
+endif()
```
As above

nordicjm commented on boards/arm/pan1783_pan1783a_evb/cpuapp_common.dtsi:
```diff
@@ -3,7 +3,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
-#include "pan1783_evb_cpuapp_common-pinctrl.dtsi"
+#include "cpuapp_common-pinctrl.dtsi"
```
let's keep the board in all the filenames (this applies to all includes and filenames) so that this requires the board prefix to be included and won't be wrongly included if someone has a same named file elsewhere, won't comment on others but please address those too

nordicjm commented on boards/arm/pan1783_pan1783a_evb/pan1783a_evb_cpunet.dts:
```diff
@@ -0,0 +1,215 @@
+
+arduino_serial: &uart0{};
+
+arduino_i2c: &i2c0 {
+	compatible = "nordic,nrf-twim";
+	/* Cannot be used together with uart0. */
+	/* status = "okay"; */
+	pinctrl-0 = <&i2c0_default>;
+	pinctrl-1 = <&i2c0_sleep>;
+	pinctrl-names = "default", "sleep";
+};
+
+arduino_spi: &spi0 {
+	compatible = "nordic,nrf-spim";
+	/* Cannot be used together with uart0. */
+	/* status = "okay"; */
+	cs-gpios = <&arduino_header 16 GPIO_ACTIVE_LOW>; /* D10 */
+	pinctrl-0 = <&spi0_default>;
+	pinctrl-1 = <&spi0_sleep>;
+	pinctrl-names = "default", "sleep";
+};
```
why would these be assigned to cpunet out of interest, and not cpuapp?

pideu-sj replied:
I did this because the network core can also have spi and i2c interfaces. When just implementing on network core, that can be helpful. But the main reason was, that Nordic also did this for network core, when looking into the nrf5340dk_nrf5340_cpunet.dts

nordicjm commented on boards/arm/pan1783_pan1783a_evb/pan1783a_evb_cpunet_reset.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2023 Panasonic Industrial Devices Europe GmbH
```
This file can go

pideu-sj commented:
@nordicjm @anangl will this work for you after doing requested changes yesterday? Thanks.

anangl commented on boards/arm/pan1783_pan1783a_evb/pan1783_evb_cpunet.dts:
Why not use a common file also in this case? The content is almost exactly the same for pan1783 and pan1783a.

anangl commented on boards/arm/pan1783_pan1783a_evb/Kconfig:
```diff
@@ -41,12 +41,12 @@ config DOMAIN_CPUNET_BOARD
 	string
-	default "pan1783_evb_cpuapp"
-	depends on BOARD_PAN1783_EVB_CPUNET
+	default "pan1783_evb_cpuapp" if BOARD_PAN1783_EVB_CPUNET
+	default "pan1783a_evb_cpuapp" if BOARD_PAN1783A_EVB_CPUNET
```
You need to use a similar construct in `config DOMAIN_CPUNET_BOARD` above.

pideu-sj commented:
@anangl @nordicjm
Did the requested changes.

nordicjm commented on boards/arm/pan1783_pan1783a_evb/doc/index.rst:
```diff
@@ -0,0 +1,62 @@
+Programming and Debugging
+*************************
+
+Please use the ``pan1783_evb_cpuapp/pan1783a_evb_cpuapp`` for application
+core or ``pan1783_evb_cpunet/pan1783_evb_cpunet`` board configuration for
```
```suggestion
Please use the ``pan1783_evb_cpuapp`` or ``pan1783a_evb_cpuapp`` for application
core and ``pan1783_evb_cpunet`` or ``pan1783_evb_cpunet`` board configuration for
```

nordicjm reviewed (changes requested):
Minor change
